DAGE-3: Implementation of search engine abstraction and class that implements it in Solr

nicolo-rinaldi opened the pull request:
## Changes
- Interface at `rre-dataset-generator/src/search_engine/interface.py`
- Code to interact with Solr search engine at `rre-dataset-generator/src/search_engine/solr_search_engine.py`
- Added tests in the `tests/integration/` folder with mocks (to see it they make sense done like that)

[Jira ticket](https://sease.atlassian.net/browse/DAGE-3)

nicolo-rinaldi committed:
[DAGE-4] minor changes
## changes
- updated README.md with new config.yaml file example
- changed slightly the Config class by making the method .load as static and not a function outside the class
- updated then the code everywhere that function is used

nicolo-rinaldi committed:
[DAGE-3] first version of the solr search engine + interface
## Changes
- added 'interface.py' (abstract class to deal with search engines) and 'solr_search_engine.py' to do search in the solr search engine
- modified file 'pyproject.toml' with request library
- added 'utils.py' file to 'src/' folder to collapse there all the general functions that might be useful in the main script

nicolo-rinaldi committed:
[DAGE-3] added test for solr search engine
## Changes
- added integration test to deal with solr with mocks; there are 2 of them: one with status code 200 and one with status code 500 (error)
- added 'resources/' folder to do tests without touching files in the main folder
- minor changes to other files (refactoring the imports)

nicolo-rinaldi committed:
[DAGE-3] changed general 'Exception' to 'ValueError' in 'test_solr_search_engine.py' file

nicolo-rinaldi committed:
[DAGE-3] Added whitespaces at the end of the python files where it was missing

nseidan commented on rre-dataset-generator/dataset_generator.py:
```diff
@@ -20,7 +11,7 @@ def parse_args():
+        config = Config.load(args.config_file)
         log.debug("Configuration loaded successfully.")
     except Exception as e:
         log.debug(f"Error loading configuration: {e}")
```
why log.debug ? not log.error if exception occurs ?

nicolo-rinaldi replied:
yeah, it makes sense. Fixed

nseidan commented on rre-dataset-generator/result.json:
```diff
@@ -0,0 +1 @@
+[]
```
There is no newline at the end, you can see git shows with the red circle

nicolo-rinaldi replied:
fixed

nseidan commented on rre-dataset-generator/src/search_engine/solr_search_engine.py:
```diff
@@ -0,0 +1,85 @@
+        if response.status_code == 200:
+            return response.json()['response']['docs']
+        else:
+            raise ValueError()
```
ValueError is very vague, better to throw an error with http code, if it is server error or client and also log error?

nicolo-rinaldi replied:
I'll add more error handling (to review) in the next commit

Intrinsical-AI replied:
I'd better add log.error and keeping ValueError() or some try-except, than defining custom Exceptions.

nseidan commented on rre-dataset-generator/tests/integration/resources/good_config.yaml:
```diff
@@ -0,0 +1,17 @@
+search_engine_collection_endpoint: "http://localhost:8983/solr/testcore"
+documents_filter:
+  - title:
+      - "and"
```
title is "and"?

nicolo-rinaldi replied:
yes, this is just to look if we have returned files

nseidan commented on rre-dataset-generator/src/search_engine/solr_search_engine.py:
```diff
@@ -0,0 +1,85 @@
+
+    def search(self, payload: Dict[str, Any]) -> List[Dict[str, Any]]:
+        """Search for documents using a query."""
+        search_url = os.path.join(str(self.endpoint), "select")
```
I just looked up `os.path.join` and it seems  it is used to construct file system path, for Windows , it uses ' \ ' backslash instead of '/', so you might end up getting something like this ` ../testcore\select`  which might lead to http not found url.

nicolo-rinaldi replied:
I hope the user working on Windows will introduce as a param in a the config file a viable Windows-like path. Am I wrong?

nseidan replied:
@nicolo-rinaldi I meant `os.path.join` is for file path construction, for http path, we should avoid using the os.path.join Also we shouldnt assume the user introduces its own param for windows os.

nseidan replied:
@nicolo-rinaldi,  take a look into `urljoin`  https://stackoverflow.com/questions/10893374/python-confusions-with-urljoin, it seems works well for constructing urls.

nicolo-rinaldi replied:
yes, that's for sure a safer option. Fixed

nseidan commented on rre-dataset-generator/src/search_engine/solr_search_engine.py:
```diff
@@ -0,0 +1,85 @@
+
+        return self.search(payload)
+
+    def extract_documents_to_evaluate_system(self, query_template: str, keyword: str="*:*") -> List[Dict[str, Any]]:
```
very long method name, maybe` fetch_for_evaluation`?

nseidan replied:
how about adding unit test for `template_to_json_body` if it correctly handles edge cases, including documents_filter?

nicolo-rinaldi replied:
done

nseidan commented on rre-dataset-generator/src/search_engine/solr_search_engine.py:
```diff
@@ -0,0 +1,85 @@
+            'params': {k: v[0] for k, v in json_body.items() if k != 'q'}
+        }
+
+    def extract_documents_to_generate_queries(self,
```
lengthy method name, maybe `fetch_for_query_generation` ?

nicolo-rinaldi replied:
Fixed

nseidan commented on rre-dataset-generator/src/search_engine/solr_search_engine.py:
```diff
@@ -0,0 +1,85 @@
+            dict: A dictionary representing the query parameters.
+        """
+        # Parse the query string into a dictionary
+        json_body = parse_qs(template_payload)
```
does this method handle quotes and a question mark?

nicolo-rinaldi replied:
I added tests and seems working as expected

nseidan commented on rre-dataset-generator/src/search_engine/solr_search_engine.py:
```diff
@@ -0,0 +1,85 @@
+
+    @staticmethod
+    def template_to_json_body(template_payload: str) -> dict:
+        """
```
add return type as well like Dict[str, Any] to be consistent in the code.

nicolo-rinaldi replied:
changed

nseidan commented on rre-dataset-generator/try_solr.py:
```diff
@@ -0,0 +1,26 @@
+
+    docs = search_engine.extract_documents_to_evaluate_system(keyword="and", query_template=config.query_template)
+
+    with open("result.json", "w") as file:
```
adding solr to file name? solr_extracted_dataset or anything else ?

nicolo-rinaldi replied:
Picked `solr_extracted_dataset.json` as name and added empty line at the end of the file

Intrinsical-AI commented on rre-dataset-generator/pyproject.toml:
I think argparse it's included by default in python std. lib for Python>=2.7 (therefore, in python 3.10). We can remove it from the toml file

nicolo-rinaldi replied:
I needed to update the dependencies and it came out. I think there are no problem if we keep the argparse dependency

dantuzi commented on rre-dataset-generator/dataset_generator.py:
```diff
@@ -20,7 +11,7 @@ def parse_args():
 
     try:
-        config = load_config(args.config_file)
+        config = Config.load(args.config_file)
```
I like it

Intrinsical-AI commented on rre-dataset-generator/src/config.py:
```diff
@@ -42,14 +44,14 @@ def check_config_type(cls, v):
-        raw_config = yaml.safe_load(f)
-    return Config(**raw_config)
\ No newline at end of file
+    @staticmethod
```
staticmethod makes a lot of sense, but, we can compare with classmethod. Why? What if tomorrow we have another config (ej: AWSConfig) that inherits from Config? -> @staticmethod will still be calling the Config.load method; which breaks heritage; if we want to AWSConfig.load().

So I'd change this for a classmethod, for long term flexibility and good practices

Intrinsical-AI replied:
Wth class method, cls it's now a parameter of load, so when we call load, it's gonna work on the current class, no matter if it's Config, or a subclass.

nicolo-rinaldi replied:
done

dantuzi commented on rre-dataset-generator/try_solr.py:
Do we need to commit this file?
If we want to commit it, it should be somewhere in the integration test folder as a temporary test before we write the proper integration tests.
Furthermore, the file `result.json` should also be moved to be a temporary integration test

nicolo-rinaldi replied:
Changed to be all inside `dataset_generator.py`

dantuzi commented on rre-dataset-generator/src/search_engine/interface.py:
```diff
@@ -0,0 +1,30 @@
+
+class BaseSearchEngine(ABC):
+    def __init__(self, endpoint: str):
+        validated = EndpointValidator(endpoint=endpoint)
```
Does the HttpUrl type already validate the endpoint? How about if the validation fails?

nicolo-rinaldi replied:
It throws a ValidationError if the class find a non-url like string. I added a test about that.

Intrinsical-AI commented on rre-dataset-generator/dataset_generator.py:
```diff
@@ -1,14 +1,5 @@
+from src.utils import parse_args
 
 if __name__ == "__main__":
     from src.logger import configure_logging
```
This doesn't make a lot of sense to me. The resource saving it's minimal.
I'd unify the logs / exception handling in all the classes, at the begining. Not in the main


**_from src.logger import configure_logging
import logging
configure_logging(level=logging.DEBUG)
log = logging.getLogger(__name__)_**

nicolo-rinaldi replied:
done

Intrinsical-AI reviewed (commented):
Small proposals

nicolo-rinaldi committed:
[DAGE-3] Updated all the requests in the PR comments
## Changes
- Added logging and better Exception handling in file 'src/search_engine/solr_search_engine.py'
- updated name of 'result.json' to 'solr_extracted_dataset.json'
- changed method in 'src/config.py' to be a classmethod instead a static one
- upgraded file 'tests/integration/test_solr_search_engine.py' with more tests
- renamed the abstract methods to be more short and readable in 'src/search_engine/interface.py'
- removed logging from main file

nseidan commented on rre-dataset-generator/src/config.py:
```diff
@@ -25,31 +33,36 @@ class Config(BaseModel):
     def check_doc_type(cls, v):
         if v is not None:
             if v.suffix[1:] != "txt":
+                log.error("queries' file must have TXT extension")
```
here .txt extension right ? can you just replace TXT with .txt ?

nicolo-rinaldi replied:
fixed

nseidan commented on rre-dataset-generator/src/search_engine/interface.py:
```diff
@@ -0,0 +1,30 @@
+from abc import ABC, abstractmethod
```
by the way, just noticed, can we give more descriptive name for this file ? not just `interface.py ` which is very generic.

nicolo-rinaldi replied:
@nseidan `search_engine_interface.py` would be better?

nseidan replied:
maybe` search_engine_base.py` as the class acts as a base for all supported engines?

nicolo-rinaldi replied:
changed

nseidan commented on rre-dataset-generator/src/config.py:
```diff
@@ -1,8 +1,16 @@
 
+from src.logger import configure_logging
+
+configure_logging(level=logging.DEBUG)
```
I'd prefer logging.INFO instead of DEBUG, @dantuzi @Intrinsical-AI WDYT ?

Intrinsical-AI replied:
I'd rather to keep all in DEBUG for now. I think it should be more atomic than INFO. 

DEBUG it's intented to support information for both: devs and clients. INFO, are intended to help only clients / users. So I'd keep debug till the end of the first sprint?

nicolo-rinaldi committed:
[DAGE-3] Minor changes requested in PR comments

nicolo-rinaldi committed:
[DAGE-3] Renamed search engine interface

nseidan commented:
@nicolo-rinaldi while browsing files,  i see `try_solr.py` and` dataset_generator.py` files. So is dataset_generator main entrance to extract dataset and call LLM service? Also why we need a separate `try_solr.py`?    I am just questioning myself, because in the work item that i am doing I need a class where I can get solr data and call LLM service for queries extraction with that data.

nicolo-rinaldi commented:
The `try_solr.py` file is there just to test the connection to the Solr instance without the use of mocking (that has been used in integration tests). And also to highlight that also ticket [DAGE-7](https://sease.atlassian.net/browse/DAGE-7) has been done

nseidan commented on rre-dataset-generator/try_solr.py:
```diff
@@ -0,0 +1,27 @@
+
+    docs = search_engine.fetch_for_evaluation(keyword="and", query_template=config.query_template)
+
+    with open("solr_extracted_dataset.json", "w") as file:
```
actually we dont need this file, instead we should put this code into dataset_generator.py, without writing into a file. Then LLMService would grab this docs  for query generation.

nicolo-rinaldi committed:
[DAGE-3] removed try_solr.py file to have all the code in the main file

nseidan commented on rre-dataset-generator/dataset_generator.py:
```diff
@@ -1,26 +1,21 @@
-        log.debug("Configuration loaded successfully.")
-    except Exception as e:
-        log.debug(f"Error loading configuration: {e}")
+    with open("solr_extracted_dataset.json", "w") as file:
```
we don't need to store the extracted docs into a separate file. It is enough to check if docs  are not empty, have values; they are already stored in search engine.  It would be redundant.

nicolo-rinaldi replied:
removed

dantuzi commented on rre-dataset-generator/solr_extracted_dataset.json:
I don't think we need to commit this

nicolo-rinaldi replied:
removed

nicolo-rinaldi committed:
[DAGE-3] removed unnecessary files and updated tests to validate the url passed to SolrSearchEngine class

nicolo-rinaldi committed:
[DAGE-3] minor change

nicolo-rinaldi committed:
[DAGE-3] minor change

Intrinsical-AI commented on rre-dataset-generator/src/config.py:
```diff
@@ -25,31 +33,36 @@ class Config(BaseModel):
     @field_validator('doc_fields')
     def check_no_empty_fields(cls, v):
         if any(not f.strip() for f in v):
+            log.error("docFields cannot contain empty strings.")
```
not mandatory, but maybe would be nice to store the strings into vars, and reuse them. Small savings on space and time!

Intrinsical-AI commented on rre-dataset-generator/src/config.py:
```diff
@@ -25,31 +33,36 @@ class Config(BaseModel):
-def load_config(config_path: str) -> Config:
-    """
-    Load and validate configuration from a YAML file.
+        :param config_path: Path to the YAML config file
```
Only redability: I would keep consistency between the Docstrings. If we use :param, we should use in all the functions, and don't mix with other formats as:
"
        Args:
            template_payload (str): The Solr query string, e.g., 'q=ghosts&fq=genre:horror&wt=json'.
        Returns:
            dict: A dictionary representing the query parameters."
            
It's not critical ofc

Intrinsical-AI commented on rre-dataset-generator/src/config.py:
```diff
@@ -25,31 +33,36 @@ class Config(BaseModel):
-    return Config(**raw_config)
\ No newline at end of file
+        log.debug("Configuration file loaded successfully.")
+        return cls(**raw_config)
```
Nice one!

Intrinsical-AI commented on rre-dataset-generator/src/search_engine/solr_search_engine.py:
```diff
@@ -0,0 +1,120 @@
+            case 200:
+                log.debug("Solr query successful.")
+                log.debug(f"URL: {search_url}\n")
+                log.debug(f"Payload: {payload}\n")
```
Redability: I'd make a single string in one line, with the 3 sentences, playing with \n if needed to better visualization

Intrinsical-AI commented on rre-dataset-generator/tests/unit/test_config.py:
```diff
@@ -1,13 +1,13 @@
 @pytest.fixture
 def config():
-    return load_config("tests/unit/resources/good_config.yaml")
+    return Config.load("tests/unit/resources/good_config.yaml")
```
Nice :+1: x2

Intrinsical-AI reviewed (approved):
Redability improvements - neither of them critical.

nicolo-rinaldi committed:
[DAGE-3] Added Document schema

## Changes
- added Document schema to work with a unified data structure
- modified the both abstraction and solr search engine class to deal with those files and added the param doc_fields so retrieve only the fields we need (from config.yaml file)
- updated pyproject.toml to resolve conflicts
- minor changes to typos
